### PR TITLE
Add initial CFR and Kuhn Poker implementation

### PR DESCRIPTION
**Pull request by chirnsch**
* Add initial CFR implementation (regular regert matching, and regret mataching plus).
* Add Kuhn Poker implementation to test algorithm.
* Add integration test for CFR with Kuhn Poker.

### CODE REVIEW

File: src/dd_cfr/algorithms/cfr.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+    def get_average_policy(self, state: str) -> Mapping[base_game.Action, float]:
+        return self._get_average(
+            self.cummulative_policies[state],
+            list(self.cummulative_policies[state].keys()),
+        )
+
+    def get_policy(self) -> Dict[str, Dict[base_game.Action, float]]:
+        policy = {}
+        for state in sorted(self.cummulative_policies.keys()):
```
Comment (DavidToneian) [inline src/dd_cfr/algorithms/cfr.py]: Why does this need to be `sorted`? Is this behavior depended upon somewhere?

File: src/dd_cfr/algorithms/cfr.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+        :param regret_matching_plus: Whether to use regret matching plus, defaults to
+            False.
+        """
+        self._cfr = CFR()
+        self._regret_matching_plus = regret_matching_plus
+
+    def _traverse(self, game: base_game.Game, reach_probs: List[float]) -> List[float]:
+        """
+        Recurisvely traverses the game tree.
```
Comment (DavidToneian) [inline src/dd_cfr/algorithms/cfr.py]: I don't care much, but I think there's a linter rule that says that docstrings should be in the imperative (i.e., `Recursively traverse [...]` instead of `traverses`). Though evidently the linter didn't find this case. :)

File: src/dd_cfr/algorithms/cfr.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+        :param game: The game to traverse.
+        :param reach_probs: The current reach probabilities.
+        :return: The expected payoffs for both players.
+        :raises ValueError: If different policy and legal_actions length encounetered.
+
```
Comment (DavidToneian) [inline src/dd_cfr/algorithms/cfr.py]: Spurious newline

File: src/dd_cfr/algorithms/cfr.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        """
+        if not game.is_terminal():
+            if game.get_active_player() == common.CHANCE_PLAYER:
```
Comment (DavidToneian) [inline src/dd_cfr/algorithms/cfr.py:132]: Instead of having `common.CHANCE_PLAYER`, maybe one should have `game.active_player_is_chance_player()`?

File: src/dd_cfr/algorithms/cfr.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+    def solve(self, game: Type[base_game.Game], iterations: int) -> None:
+        """
+        Solve a nash equilibrium for the provided game.
+
+        :param game: The game to solve.
+        :param iterations: Number of traversals.
+        """
+        for _ in range(iterations):
+            self._traverse(game(), [1, 1, 1])
```
Comment (DavidToneian) [inline src/dd_cfr/algorithms/cfr.py]: `[1, 1, 1]` is hard-coded for Kuhn poker I suppose?

File: src/dd_cfr/games/base_game.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        """
+
+    @abc.abstractmethod
+    def child(self, action: Action) -> Game:
+        """
+        Returns a copy of the current game state with the given action applied.
+        """
+
+    def _get_other_player(self, player: int) -> int:
+        return (player + 1) % 2
```
Comment (DavidToneian) [inline src/dd_cfr/games/base_game.py:50]: Assumes a two-player game (and that the chance player is not a player).

Let's maybe have `Game`s declare their number of players, and `assert` here that it's `2`?

File: src/dd_cfr/games/kuhn_poker.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+        cards: Optional[List[ChanceAction]] = None,
+        history: Optional[List[PlayerAction]] = None,
+    ) -> None:
+        """
+        Initialize KuhnPoker class
+
+        :param cards: Optional current cards, defaults to None.
+        :param history: Optional current history, defaults to None.
+        """
+        if cards is None:
```
Comment (DavidToneian) [inline src/dd_cfr/games/kuhn_poker.py]: Shorter alternative, though I don't have strong feelings either way:

```
self._cards = cards or []
self._history = history or []
```

File: src/dd_cfr/algorithms/cfr.py [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+                legal_actions = game.get_legal_actions()
+                policy = self._cfr.get_current_policy(game.get_state(), legal_actions)
+
+            rewards = {}
+            for action, probability in policy.items():
+                next_reach_probs = reach_probs[:]
+                next_reach_probs[game.get_active_player()] *= probability
+                rewards[action] = self._traverse(game.child(action), next_reach_probs)
+
+            payoffs = [0.0, 0.0]
```
Comment (DavidToneian) [inline src/dd_cfr/algorithms/cfr.py:144]: Assumes a two-player game (and that the chance player is not a player).

Let's maybe have `Game`s declare their number of players, and `assert` here that it's `2`?

(Same below for `range(2)`, `get_inactive_player`.)

File: src/dd_cfr/games/kuhn_poker.py [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+
+    def _get_formatted_card(self, card: ChanceAction) -> str:
+        return card.name
+
+    def _get_formatted_history(self) -> str:
+        return ", ".join(str(pa.action.name) for pa in self._history)
+
+    def get_state(self) -> str:
+        if len(self._cards) < 2:  # pragma: no cover
+            return ""
```
Comment (DavidToneian) [inline src/dd_cfr/games/kuhn_poker.py:77]: Would this be an error state? If yes, why not just `assert len(self._cards) >= 2`? If no, why the `pragma: no cover`?

File: src/dd_cfr/games/kuhn_poker.py [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+        payoffs[winner] = winning_amount
+        payoffs[self._get_other_player(winner)] = -winning_amount
+
+        return payoffs
+
+    def get_legal_actions(self) -> Sequence[Action]:
+        if not self._history or self._history[-1].action == Action.CHECK:
+            return [Action.CHECK, Action.BET]
+
+        if self._history[-1].action == Action.BET:
```
Comment (DavidToneian) [inline src/dd_cfr/games/kuhn_poker.py:117]: Let's make this an `assert`, and save the `raise` with `pragma: no cover` below.

File: src/dd_cfr/games/kuhn_poker.py [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+        raise ValueError(
+            f"Should not reach this state after {self._history[-1]}"
+        )  # pragma: no cover
+
+    def get_chance_probabilities(self) -> Mapping[base_game.Action, float]:
+        if self.get_active_player() != common.CHANCE_PLAYER:
+            raise ValueError(
+                "Should only call get_chance_probabilities when the chance player is"
+                " active."
+            )  # pragma: no cover
```
Comment (DavidToneian) [inline src/dd_cfr/games/kuhn_poker.py:134]: This could be covered in a test, or treated with an `assert`.

File: src/dd_cfr/algorithms/cfr.py (hunk 1 above)
Comment (chirnsch) [reply src/dd_cfr/algorithms/cfr.py]: Sorting is not needed, removed. I'm sorting in the printing function to make the order deterministic, but no need to sort here (artifact from previous implementation).

File: src/dd_cfr/algorithms/cfr.py (hunk 5 above)
Comment (chirnsch) [reply src/dd_cfr/algorithms/cfr.py]: Added more documentation and used a tuple instead of a list to make the type annotation clearer. The reach probabilities are for player 1, player 2, and the chance player - which should be appliciaple for all games. At the start of the game, the reach probabilities are `1` for all players.

File: src/dd_cfr/algorithms/cfr.py (hunk 4 above)
Comment (chirnsch) [reply src/dd_cfr/algorithms/cfr.py:132]: I agree this is not great. Since we occassionally might want to show/print the active player as well, I find it nice to have a common way of representing players across games.

It might best might actually be to have a Player ENUM for Player 1, Player 2, and the chance player, with some functionality to e.g. get the other non-chance player (which I currently added in base_game.py). What do you think?

Added a TODO in common.py for this for now..

File: src/dd_cfr/algorithms/cfr.py (hunk 3 above)
Comment (chirnsch) [reply src/dd_cfr/algorithms/cfr.py]: Thanks, done.

File: src/dd_cfr/games/kuhn_poker.py (hunk 7 above)
Comment (chirnsch) [reply src/dd_cfr/games/kuhn_poker.py]: This would also create new empty lists if cards or history is empty, which is not an issue but I don't like conceptually.

I chose a middleground one-layer instead.

File: src/dd_cfr/games/kuhn_poker.py (hunk 10 above)
Comment (chirnsch) [reply src/dd_cfr/games/kuhn_poker.py:117]: One of the linters (I believe) complained about the usage of assert, since it can be disabled in production. Would you prefer to change the linter behavior and switch to assert?

File: src/dd_cfr/games/kuhn_poker.py (hunk 11 above)
Comment (chirnsch) [reply src/dd_cfr/games/kuhn_poker.py:134]: (same as above).

File: src/dd_cfr/algorithms/cfr.py (hunk 8 above)
Comment (chirnsch) [reply src/dd_cfr/algorithms/cfr.py:144]: Let's discuss offline whether to restrict oureselves to two player games only.

File: src/dd_cfr/games/kuhn_poker.py (hunk 9 above)
Comment (chirnsch) [reply src/dd_cfr/games/kuhn_poker.py:77]: No, that would not be an error - updated the logic and added a comment to make it clear.

File: src/dd_cfr/algorithms/cfr.py [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+from dd_cfr.games import base_game
+
+
+class CFR:
+    """CFR class."""
+
+    def __init__(self) -> None:
+        """Initialize CFR class."""
+        # state, action to regret, to compute current policy.
+        self.cummulative_regrets: dict[
```
Comment (DavidToneian) [inline src/dd_cfr/algorithms/cfr.py]: `cumulative` with just one `m`

File: src/dd_cfr/algorithms/cfr.py [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+from dd_cfr import common
+from dd_cfr.games import base_game
+
+
+class CFR:
+    """CFR class."""
+
+    def __init__(self) -> None:
+        """Initialize CFR class."""
+        # state, action to regret, to compute current policy.
```
Comment (DavidToneian) [inline src/dd_cfr/algorithms/cfr.py]: Somewhat hard to parse. Is `# Maps state and action to regret. Used to compute current policy` correct and more understandable? If so, analogous change in line 22 below.

File: src/dd_cfr/algorithms/cfr.py [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+
+
+class CFRSolver:
+    """CFR Solver, traverses the provided game to compute a nash equilibrium."""
+
+    def __init__(self, regret_matching_plus: bool = False) -> None:
+        """Initialize CFRSolver class.
+
+        :param regret_matching_plus: Whether to use regret matching plus, defaults to
+            False.
```
Comment (DavidToneian) [inline src/dd_cfr/algorithms/cfr.py]: Refer to https://arxiv.org/abs/1407.5042 again (and use `regret-matching+` if that's the nomenclature they use)?

File: src/dd_cfr/algorithms/cfr.py [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+        game: base_game.Game,
+        reach_probs: Sequence[float] = (1.0, 1.0, 1.0),
+    ) -> Sequence[float]:
+        """Recurisvely traverse the game tree.
+
+        :param game: The game to traverse.
+        :param reach_probs: The current reach probabilities for player 1, player 2 and
+            the chance player.
+        :return: The expected payoffs for both players.
+        """
```
Comment (DavidToneian) [inline src/dd_cfr/algorithms/cfr.py:130]: Maybe as a todo for the future, but we could check here whether `reach_probs` is of length `3`.

File: src/dd_cfr/games/base_game.py [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+    def get_state(self) -> str:
+        """Return the state from the perspective of the currently active player."""
+
+    @abc.abstractmethod
+    def is_terminal(self) -> bool:
+        """Return whether the current state is terminal."""
+
+    @abc.abstractmethod
+    def get_payoffs(self) -> list[float]:
+        """Return the payoffs for all players in order."""
```
Comment (DavidToneian) [inline src/dd_cfr/games/base_game.py]: Somewhat unclear whether "in order" means "for players 0, 1, ..., in that order", or "in the order of play, starting with the active player."

File: src/dd_cfr/games/kuhn_poker.py [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+
+        payoffs[winner] = winning_amount
+        payoffs[self._get_other_player(winner)] = -winning_amount
+
+        return payoffs
+
+    def get_legal_actions(self) -> Sequence[Action]:
+        """Return the legal actions for the active (possibly chance) player.
+
+        :raises ValueError: If legal actions are retrieved for an  impossible state.
```
Comment (DavidToneian) [inline src/dd_cfr/games/kuhn_poker.py]: Nit: two spaces before `impossible`

**Review by DavidToneian — CHANGES_REQUESTED** [review]

File: src/dd_cfr/algorithms/cfr.py (hunk 12 above)
Comment (chirnsch) [reply src/dd_cfr/algorithms/cfr.py]: Thanks, done.

File: src/dd_cfr/algorithms/cfr.py (hunk 14 above)
Comment (chirnsch) [reply src/dd_cfr/algorithms/cfr.py]: Done.

File: src/dd_cfr/games/base_game.py (hunk 16 above)
Comment (chirnsch) [reply src/dd_cfr/games/base_game.py]: Done (the former option).

File: src/dd_cfr/games/base_game.py (hunk 6 above)
Comment (chirnsch) [reply src/dd_cfr/games/base_game.py:50]: Discussed offline, we focus on 2 player games for now.

File: src/dd_cfr/games/kuhn_poker.py (hunk 17 above)
Comment (chirnsch) [reply src/dd_cfr/games/kuhn_poker.py]: Thanks, done.

File: src/dd_cfr/algorithms/cfr.py (hunk 13 above)
Comment (chirnsch) [reply src/dd_cfr/algorithms/cfr.py]: Thanks, updated to your suggestion.

File: src/dd_cfr/algorithms/cfr.py (hunk 15 above)
Comment (chirnsch) [reply src/dd_cfr/algorithms/cfr.py:130]: I'd leave this as is - there are many things we could asset as well, e.g. that all probabilities are <= 1, but I wouldn't check all of them exhaustively (at least for now).

**Review by DavidToneian — APPROVED** [review]
